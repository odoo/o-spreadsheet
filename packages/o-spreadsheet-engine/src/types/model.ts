import { GeoChartRegion } from "./chart/geo_chart";
import { Client, ClientPosition } from "./collaborative/session";
import { TransportService } from "./collaborative/transport_service";
import { Currency } from "./currency";
import { InformationNotification } from "./env";
import { FileStore } from "./files";
import { Locale } from "./locale";
import { Color } from "./misc";

/**
 * Model
 *
 * The Model class is the owner of the state of the Spreadsheet. However, it
 * has more a coordination role: it defers the actual state manipulation work to
 * plugins.
 *
 * At creation, the Model instantiates all necessary plugins. They each have
 * a private state (for example, the Selection plugin has the current selection).
 *
 * State changes are then performed through commands.  Commands are dispatched
 * to the model, which will then relay them to each plugins (and the history
 * handler). Then, the model will trigger an 'update' event to notify whoever
 * is concerned that the command was applied (if it was not cancelled).
 *
 * Also, the model has an unconventional responsibility: it actually renders the
 * visible viewport on a canvas. This is because each plugins actually manage a
 * specific concern about the content of the spreadsheet, and it is more natural
 * if they are able to read data from their internal state to represent it on the
 * screen.
 *
 * Note that the Model can be used in a standalone way to manipulate
 * programmatically a spreadsheet.
 */
export type Mode = "normal" | "readonly" | "dashboard";

export interface ModelConfig {
  readonly mode: Mode;
  /**
   * Any external custom dependencies your custom plugins or functions might need.
   * They are available in plugins config and functions
   * evaluation context.
   */
  readonly custom: Readonly<{
    [key: string]: any;
  }>;
  readonly defaultCurrency?: Partial<Currency>;
  /**
   * External dependencies required to enable some features
   * such as uploading images.
   */
  readonly external: Readonly<ModelExternalConfig>;
  readonly moveClient: (position: ClientPosition) => void;
  readonly transportService: TransportService;
  readonly client: Client;
  readonly snapshotRequested: boolean;
  readonly notifyUI: (payload: InformationNotification) => void;
  readonly raiseBlockingErrorUI: (text: string) => void;
  readonly customColors: Color[];
}

export interface ModelExternalConfig {
  readonly fileStore?: FileStore;
  readonly loadCurrencies?: () => Promise<Currency[]>;
  readonly loadLocales?: () => Promise<Locale[]>;
  readonly geoJsonService?: {
    getAvailableRegions: () => GeoChartRegion[];
    getTopoJson: (region: string) => Promise<any>;
    /**  Convert the name of a geographical feature (eg. France) to the id of the corresponding feature in the TopoJSON */
    geoFeatureNameToId: (region: string, territory: string) => string | undefined;
  };
}
