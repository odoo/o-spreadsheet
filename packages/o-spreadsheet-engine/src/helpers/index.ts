export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
