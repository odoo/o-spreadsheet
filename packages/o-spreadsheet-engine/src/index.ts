export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
