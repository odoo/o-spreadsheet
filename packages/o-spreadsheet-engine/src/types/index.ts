export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
