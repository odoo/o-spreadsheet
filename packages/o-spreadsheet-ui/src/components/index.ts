export { Spreadsheet } from "./spreadsheet/spreadsheet";
