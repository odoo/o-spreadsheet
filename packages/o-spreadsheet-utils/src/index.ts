export * from "./translation";
export * from "./uuid";
