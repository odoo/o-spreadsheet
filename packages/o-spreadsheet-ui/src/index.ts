import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}