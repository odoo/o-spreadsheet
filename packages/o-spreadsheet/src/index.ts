import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };