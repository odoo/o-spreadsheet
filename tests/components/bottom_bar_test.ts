import { Component, hooks, tags } from "@odoo/owl";
import { BottomBar } from "../../src/components/bottom_bar";
import { Model } from "../../src/model";
import { makeTestFixture, nextTick } from "../helpers";
import { triggerMouseEvent } from "../dom_helper";
import { CommandResult } from "../../src/types";

const { xml } = tags;
const { useSubEnv } = hooks;

let fixture: HTMLElement;

class Parent extends Component<any, any> {
  static template = xml`<BottomBar model="model"/>`;
  static components = { BottomBar };
  model: Model;
  constructor(model: Model) {
    super();
    useSubEnv({
      openSidePanel: (panel: string) => {},
      dispatch: jest.fn(() => ({ status: "SUCCESS" } as CommandResult)),
      getters: model.getters,
      askConfirmation: jest.fn(),
    });
    this.model = model;
  }
  mounted() {
    this.model.on("update", this, this.render);
  }

  willUnmount() {
    this.model.off("update", this);
  }
}

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

describe("BottomBar component", () => {
  test("simple rendering", async () => {
    const parent = new Parent(new Model());
    await parent.mount(fixture);

    expect(fixture.querySelector(".o-spreadsheet-bottom-bar")).toMatchSnapshot();
  });

  test("Can create a new sheet", async () => {
    const parent = new Parent(new Model());
    await parent.mount(fixture);

    triggerMouseEvent(".o-add-sheet", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("CREATE_SHEET", { activate: true });
  });

  test("Can activate a sheet", async () => {
    const parent = new Parent(new Model());
    await parent.mount(fixture);

    triggerMouseEvent(".o-sheet", "click");
    const from = parent.model.getters.getActiveSheet();
    const to = from;
    expect(parent.env.dispatch).toHaveBeenCalledWith("ACTIVATE_SHEET", { from, to });
  });

  test.skip("Can open context menu of a sheet", async () => {
    const parent = new Parent(new Model());
    await parent.mount(fixture);

    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
  });
});
