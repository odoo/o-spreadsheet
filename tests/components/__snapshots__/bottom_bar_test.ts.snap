// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBar component simple rendering 1`] = `
<div
  class="o-spreadsheet-bottom-bar"
>
  <div
    class="o-sheet-item o-add-sheet"
  >
    <svg
      class="o-icon"
    >
      <path
        d="M8,0 L10,0 L10,8 L18,8 L18,10 L10,10 L10,18 L8,18 L8,10 L0,10 L0,8 L8,8"
        fill="#000000"
      />
    </svg>
  </div>
  <div
    class="o-sheet-item o-list-sheets"
  >
    <svg
      class="o-icon"
      viewBox="0 0 384 384"
    >
      <rect
        height="42.667"
        width="384"
        x="0"
        y="277.333"
      />
      <rect
        height="42.667"
        width="384"
        x="0"
        y="170.667"
      />
      <rect
        height="42.667"
        width="384"
        x="0"
        y="64"
      />
    </svg>
  </div>
  <div
    class="o-all-sheets"
  >
    <div
      class="o-sheet-item o-sheet active"
      data-id="Sheet1"
      title="Sheet1"
    >
      <span
        class="o-sheet-name"
      >
        Sheet1
      </span>
      <span
        class="o-sheet-icon"
      >
        <svg
          class="o-icon"
        >
          <polygon
            fill="#000000"
            points="0 0 4 4 8 0"
            transform="translate(5 7)"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
